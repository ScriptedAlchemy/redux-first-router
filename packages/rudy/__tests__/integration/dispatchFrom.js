import createTest from '../../__helpers__/createTest'

createTest(
  're-dispatching state.from behaves as if original action was dispatched',
  {
    SECOND: {
      path: '/second',
      beforeEnter: async ({ dispatch, action }) => {
        if (action.state.allow) return
        return dispatch({ type: 'REDIRECTED' })
      },
    },
  },
  [],
  async ({ dispatch, getState }) => {
    let res = await dispatch({ type: 'SECOND' })

    expect(res.type).toEqual('REDIRECTED_COMPLETE')

    // users can simply dispatch `state.from` whenever they are ready to take the user
    // to where the user initially tried to go
    const { from } = getState().location
    expect(from).toMatchSnapshot()
    from.state = { allow: true } // for our simple example, we'll add this key/val to make `beforeEnter` not redirect
    res = await dispatch(from)

    expect(res.type).toEqual('SECOND')

    expect(res).toMatchSnapshot('response')
    expect(getState()).toMatchSnapshot('state')
  },
)
