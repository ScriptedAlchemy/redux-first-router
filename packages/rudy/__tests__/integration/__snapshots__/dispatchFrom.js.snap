// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`re-dispatching state.from behaves as if original action was dispatched 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "index": 1,
    "key": "345678",
    "pathname": "/second",
    "scene": "",
    "search": "",
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "SECOND",
}
`;

exports[`re-dispatching state.from behaves as if original action was dispatched: response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/first",
          "scene": "",
          "search": "",
          "url": "/first",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {
          "allow": true,
        },
        "type": "SECOND",
      },
    ],
    "from": null,
    "index": 2,
    "key": "345678",
    "kind": "push",
    "length": 3,
    "n": 1,
    "pathname": "/second",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/redirected",
        "scene": "",
        "search": "",
        "url": "/redirected",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "REDIRECTED",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {
    "allow": true,
  },
  "type": "SECOND",
}
`;

exports[`re-dispatching state.from behaves as if original action was dispatched: state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/first",
          "scene": "",
          "search": "",
          "url": "/first",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {
          "allow": true,
        },
        "type": "SECOND",
      },
    ],
    "from": null,
    "hash": "",
    "index": 2,
    "key": "345678",
    "kind": "push",
    "length": 3,
    "n": 1,
    "params": Object {},
    "pathname": "/second",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/redirected",
        "scene": "",
        "search": "",
        "url": "/redirected",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "REDIRECTED",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "allow": true,
    },
    "status": 200,
    "type": "SECOND",
    "universal": false,
    "url": "/second",
  },
  "title": "SECOND",
}
`;
